src/contracts/core/AllocationManager.sol

```diff
@@ -0,0 +1,786 @@
+     *
+     */
+
+    /**
+     * @dev Initializes the DelegationManager address, the deallocation delay, and the allocation configuration delay.
+     */
+    constructor(
+        IDelegationManager _delegation,
+        IPauserRegistry _pauserRegistry,
+        uint32 _DEALLOCATION_DELAY,
+        uint32 _ALLOCATION_CONFIGURATION_DELAY
+    )
+        AllocationManagerStorage(_delegation, _DEALLOCATION_DELAY, _ALLOCATION_CONFIGURATION_DELAY)
+        Pausable(_pauserRegistry)
+    {
+        _disableInitializers();
+    }
+
+    /// @inheritdoc IAllocationManager
+    function initialize(address initialOwner, uint256 initialPausedStatus) external initializer {
+        _setPausedStatus(initialPausedStatus);
+        _transferOwnership(initialOwner);
+    }
+
+    /// @inheritdoc IAllocationManager
+    function slashOperator(
+        SlashingParams calldata params
+    ) external onlyWhenNotPaused(PAUSED_OPERATOR_SLASHING) {
+        require(0 < params.wadToSlash && params.wadToSlash <= WAD, InvalidWadToSlash());
+
+        // Check that the operator set exists and the operator is registered to it
+        OperatorSet memory operatorSet = OperatorSet(msg.sender, params.operatorSetId);
+        bool isRegistered = _isRegistered(params.operator, operatorSet);
+        require(_operatorSets[operatorSet.avs].contains(operatorSet.id), InvalidOperatorSet());
+        require(isRegistered, NotMemberOfSet());
+
+        uint256 length = _operatorSetStrategies[operatorSet.key()].length();
+        IStrategy[] memory strategiesSlashed = new IStrategy[](length);
+        uint256[] memory wadSlashed = new uint256[](length);
+
+        // For each strategy in the operator set, slash any existing allocation
+        for (uint256 i = 0; i < length; i++) {
+            // 1. Get the operator's allocation info for the strategy and operator set
+            IStrategy strategy = IStrategy(_operatorSetStrategies[operatorSet.key()].at(i));
+            (StrategyInfo memory info, Allocation memory allocation) =
+                _getUpdatedAllocation(params.operator, operatorSet.key(), strategy);
+            strategiesSlashed[i] = strategy;
+
+            // 2. Skip if the operator does not have a slashable allocation
+            // NOTE: this "if" is equivalent to: `if (!_isAllocationSlashable)`, because the other
+            // conditions in this method are already true (isRegistered + operatorSetStrategies.contains)
+            if (allocation.currentMagnitude == 0) {
+                continue;
+            }
+
+            // 3. Calculate the amount of magnitude being slashed, and subtract from
+            // the operator's currently-allocated magnitude, as well as the strategy's
+            // max and encumbered magnitudes
+            uint64 slashedMagnitude = uint64(uint256(allocation.currentMagnitude).mulWadRoundUp(params.wadToSlash));
+            uint256 sharesWadSlashed = uint256(slashedMagnitude).divWad(info.maxMagnitude);
```

Comment:
should `sharesWadSlashed` also be rounded up? or is this not rounded up to ensure that delegated shares always >= sum of staker shares ?

Reply:
`sharesWadSlashed` is not as important in terms of rounding since it's just passed into the event. The `DM.burnOperatorShares` call has been changed back to use `prevMaxMagnitude` and `newMaxMagnitude`. That being said, yes we do round down the amount of shares to decrement from operatorShares to ensure that property you outlined. Here's the excerpt in the SlashingLib

```
function calcSlashedAmount(
        uint256 operatorShares,
        uint256 prevMaxMagnitude,
        uint256 newMaxMagnitude
    ) internal pure returns (uint256) {
        // round up mulDiv so we don't overslash
        return operatorShares - operatorShares.mulDiv(newMaxMagnitude, prevMaxMagnitude, Math.Rounding.Up);
    }